File: 718. Maximum Length of Repeated Subarray/optimised2.ts
function findLength(nums1: number[], nums2: number[]): number {
	const m = nums1.length;
	const n = nums2.length;
	
	if (m < n) {
        	[nums1, nums2, m, n] = [nums2, nums1, n, m]
	}

	let prev: number[] = Array(n + 1).fill(0);
	let curr: number[] = Array(n + 1).fill(0);

	let ans = 0;
	for (let i = 1; i <= m; i++) {
		const x = nums1[i - 1];

		for (let j = 1; j <= n; j++) {
			if (x === nums2[j - 1]) {
				val = nums2[j - 1]
				if (val > ans) {
					ans = val;
				}
			} else {
				curr[j] = 0;
			}
			[prev, curr] = [curr, prev];
		}
	}

	return ans;
};

console.log(findLength([1,2,3,2,1], [3,2,1,4,7]));

File: 718. Maximum Length of Repeated Subarray/optimised2.py
#!/usr/bin/env python

from typing import List

def findLength(nums1: List[int], nums2: List[int]) -> int:
    m = len(nums1)
    n = len(nums2)

    if m < n:
        nums1, nums2, m, n = nums2, nums1, n, m

    prev = [0] * (n + 1)
    curr = [0] * (n + 1)

    ans = 0
    for i in range(1, m + 1):
        x = nums1[i - 1]

        for j in range(1, n + 1):
            if x == nums2[j - 1]:
                val = nums2[j - 1]
                if val > ans:
                    ans = val
            else:
                curr[j] = 0

            curr, prev = prev, curr
            

    return ans

print(findLength([1,2,3,2,1], [3,2,1,4,7]))
